mac/Keyman4MacIM/Keyman4MacIM/KMInputMethodAppDelegate.m

```diff
@@ -760,90 +760,174 @@ - (void)setContextBuffer:(NSMutableString *)contextBuffer {
-            if (!infoDict)
-                continue;
-            //NSString *str = [NSString stringWithFormat:@"%@ (%@)", [infoDict objectForKey:kKMKeyboardNameKey], [infoDict objectForKey:kKMKeyboardVersionKey]];
-            NSString *str = [infoDict objectForKey:kKMKeyboardNameKey];
-            NSMenuItem *item = [[NSMenuItem alloc] initWithTitle:str action:@selector(menuAction:) keyEquivalent:@""];
-            [item setTag:itag++];
-            if ([path isEqualToString:self.selectedKeyboard]) {
-                [item setState:NSOnState];
-                KMXFile *kmx = [[KMXFile alloc] initWithFilePath:path];
-                [self setKmx:kmx];
-                NSDictionary *kmxInfo = [KMXFile keyboardInfoFromKmxFile:path];
-                NSString *kvkFilename = [kmxInfo objectForKey:kKMVisualKeyboardKey];
-                if (kvkFilename != nil) {
-                    NSString *kvkFilePath = [self kvkFilePathFromFilename:kvkFilename];
-                    if (kvkFilePath != nil)
-                        kvk = [[KVKFile alloc] initWithFilePath:kvkFilePath];
-                }
-                [self setKvk:kvk];
-                [self setKeyboardName:[kmxInfo objectForKey:kKMKeyboardNameKey]];
-                [self setKeyboardIcon:[kmxInfo objectForKey:kKMKeyboardIconKey]];
-                [self loadSavedStores];
+    [self setDefaultKeymanMenuItems];
+    [self updateKeyboardMenuItems];
+}
 
-                didSetKeyboard = YES;
-            }
-            else
-                [item setState:NSOffState];
+- (void)setDefaultKeymanMenuItems {
+  NSMenuItem *config = [self.menu itemWithTag:CONFIG_MENUITEM_TAG];
+  if (config) {
+    [config setAction:@selector(menuAction:)];
+  }
+  
+  NSMenuItem *osk = [self.menu itemWithTag:OSK_MENUITEM_TAG];
+  if (osk) {
+    [osk setAction:@selector(menuAction:)];
+  }
+  
+  NSMenuItem *about = [self.menu itemWithTag:ABOUT_MENUITEM_TAG];
+  if (about) {
+    [about setAction:@selector(menuAction:)];
+  }
+}
 
-            [keyboards.submenu addItem:item];
+- (void)updateKeyboardMenuItems {
+  self.numberOfKeyboardMenuItems = [self calculateNumberOfKeyboardMenuItems];
+  [self removeDynamicKeyboardMenuItems];
+  [self addDynamicKeyboardMenuItems];
+}
+
+- (int)calculateNumberOfKeyboardMenuItems {
+  if (self.activeKeyboards.count == 0) {
+    // if there are no active keyboards, then we will insert one placeholder menu item 'No Active Keyboards'
+    return 1;
+  } else {
+    return (int) self.activeKeyboards.count;
+  }
+}
+
+- (void)removeDynamicKeyboardMenuItems {
+  int numberToRemove = (int) self.menu.numberOfItems - DEFAULT_KEYMAN_MENU_ITEM_COUNT;
+  
+  if (self.debugMode) {
+    NSLog(@"*** removeDynamicKeyboardMenuItems, self.menu.numberOfItems = %ld, number of items to remove = %d", (long)self.menu.numberOfItems, numberToRemove);
+  }
+  
+  if (numberToRemove > 0) {
+    for (int i = 1; i <= numberToRemove; i++) {
```

Comment:
Can we use a zero-based loop? It's more consistent with all the rest of our loops in our code

```suggestion
    for (int i = 0; i < numberToRemove; i++) {
```